keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,302 @@
+    - [Infrastructure Needed](#infrastructure-needed)
+
+
+## Summary
+
+The Kubernetes [Storage
+SIG](https://github.com/kubernetes/community/tree/master/sig-storage)
+maintains a set of components under the
+[kubernetes-csi](https://github.com/kubernetes-csi) GitHub
+organization. Those components are intentionally not part of core
+Kubernetes even though they are maintained by the Kubernetes project.
+
+This document explains how these components are released.
+
+## Motivation
+
+So far, the process for tagging and building components has been
+fairly manual, with some assistance by Travis CI. There has been no
+automatic testing on a Kubernetes cluster, neither stand-alone
+deployment of individual components nor full end-to-end (E2E) testing
+of several Kubernetes-CSI components.
+
+With CSI support reaching GA in Kubernetes 1.13, it is time to define
+and follow a less work-intensive and more predictable approach.
+
+### Goals
+
+- define the release process for each component
+- unit testing as mandatory pre-submit check for all components
+- testing on Kubernetes as mandatory pre-submit check for components that get
+  deployed in a cluster
+
+### Non-Goals
+
+- a combined "Kubernetes-CSI release": each component gets released
+  separately.  It is the responsibility of a CSI driver maintainer
+  pick and test sidecar releases for a combined deployment of that
+  driver. The [hostpath deployment
+  example](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy)
+  can be used as a starting point, but it's not going to be able to
+  test all the possible combinations of features that a specific CSI
+  driver may need.
+- change responsibilities among the kubernetes-csi maintainers: as before, each
+  component will have some main maintainer who is responsible for releasing updates
+  of that component
+- automatically generate release notes: this will be covered by a separate, future KEP
+
+
+## Proposal
+
+### Versionioning
+
+Each of the components has its own documentation, versioning and
+release notes. [Semantic versioning](https://semver.org/) with `v`
+prefix (i.e. `v1.2.3`) is used for components that provide a stable
+API, like for example:
+- external-attacher
+- external-provisioner
+- external-snapshotter
+- cluster-driver-registrar
+- node-driver-registrar
+- csi-driver-host-path
+- csi-lib-utils
+
+Other components are internal utility packages that are
+not getting tagged:
+- csi-release-tools
+
+An additional suffix like `-rc1` can be used to denote a pre-release.
+
+The [hostpath example
+deployment](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy)
+defines the exact release of each sidecar that is part of that
+deployment. When updating to newer sidecar releases, a new hostpath
+driver release is tagged with version numbers bumped according to the
+same rules as the individual components (major version bump when any
+of the sidecars had a major change, etc.).
```

Comment:
Generally, this sounds good, but there is a weird artefact not covered here...

That of the deployment objects for hostpath. they source from multiple images (hostpath and sidecars), so just using the hostpath's semver isn't enough as it can be the same but a sidecar can get updated.

I can think of two solutions currently.
1. version the deployment scripts separately
2. bump the host path version even though it doesn't actually change when deployment scripts change.

Reply:
one drawback to option two might be for those that mirror containers, they would have to mirror a new version that isn't really different then the old version just to bump the deployment scripts version. may not apply to hostpath in particular, but other future drivers maybe.

Reply:
Option 2 is what I was trying (and obviously failing :frowning:) to describe here. "When updating to newer sidecar releases, a new hostpath driver release is tagged" is exactly that: as there is no explicit version encoded in the repo, setting the tag triggers a new build of the hostpath driver with a new version that has the same source code as the previous version. The only difference in the resulting binary will be the embedded version number (a combination of tag and commit hash).

Option 1 isn't possible as long as the deployment scripts are in the same repo as the hostpath driver source code, because versioning is per repo, not per files or directory in that repo.

I'm not a fan of this arrangement, it would be cleaner to have the deployment example elsewhere. But "elsewhere" isn't that obvious either and comes with additional overhead (yet another repo to manage), so I'm fine with the proposed solution as a compromise. The hostpath driver simply is special, so it's okay to do things a bit differently than for other drivers.

Reply:
I think it makes sense to bundle deployment into a release. The integration of sidecars and driver is what you end up packaging, validating and delivering to users.

They're technically all individual containers but individually they're not very useful on their own (at least for running in kubernetes)

Reply:
While it is a special case, there will be less special cases in the immediate future that will trip across it, again. We have the repo created for the image driver and I'm about to push a pr for it. I was hoping to add release support to it soon, then saw this was cooking.

Reply:
I also added a new entry for this under "alternatives".